flight/PiOS/STM32F4xx/pios_ws2811.c

```diff
@@ -0,0 +1,386 @@
+/**
+ ******************************************************************************
+ * @addtogroup PIOS PIOS Core hardware abstraction layer
+ * @{
+ * @addtogroup PIOS_WS2811 Driver for WS2811 LEDs.
+ * @brief Driver for WS2811 LEDs.  Uses one timer (generally TIM1
+ * advanced control timer) and two DMAs.  First DMA sets the bit high at
+ * the beginning of each cycle.  Next DMA picks one of two phases to lower
+ * the bit at.
+ * @{
+ *
+ * @file       pios_ws2811.c
+ * @author     dRonin, http://dRonin.org, Copyright (C) 2016.
+ * @see        The GNU Public License (GPL) Version 3
+ *
+ *****************************************************************************/
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ *
+ * Additional note on redistribution: The copyright and license notices above
+ * must be maintained in each individual source file that is a derivative work
+ * of this source file; otherwise redistribution is prohibited.
+ */
+
+#include "pios_config.h"
+
+#if defined(PIOS_INCLUDE_WS2811)
+
+#include "pios.h"
+#include "openpilot.h"
+#include <pios_stm32.h>
+#include "stm32f4xx_tim.h"
+#include "pios_tim_priv.h"
+
+#include "pios_ws2811.h"
+
+struct ws2811_pixel_data_s {
+	uint8_t g;
+	uint8_t r;
+	uint8_t b;
+};
+
+// Each DMA side is 6 LEDs, 24 bits per pixel, 2 bytes per bit, for 288
```

Comment:
Yah.  I picked 6 because if people are using more, I need to manage the interrupt rate, as well as the amount of work during each interrupt.  This results in about a 3KHz interrupt rate during blitting out the pixel data, having to fill 144 bytes in an array on each one.